src/QMCHamiltonians/QMCHamiltonian.cpp

```diff
@@ -572,34 +560,22 @@ QMCHamiltonian::FullPrecRealType QMCHamiltonian::evaluateDeterministic(ParticleS
   for (int i = 0; i < H.size(); ++i)
   {
     ScopedTimer h_timer(my_timers_[i]);
-    const auto LocalEnergyComponent = H[i]->evaluateDeterministic(P);
-    if (qmcplusplus::isnan(LocalEnergyComponent))
-    {
-      std::ostringstream msg;
-      msg << "QMCHamiltonian::evaluateDeterministic component " << H[i]->getName() << " returns NaN." << std::endl;
-      P.print(msg);
-      throw std::runtime_error(msg.str());
-    }
-    LocalEnergy += LocalEnergyComponent;
-    H[i]->setObservables(Observables);
+    H[i]->evaluateDeterministic(P);
+    updateComponent(*H[i], *this, P);
 #if !defined(REMOVE_TRACEMANAGER)
     H[i]->collectScalarTraces();
 #endif
-    H[i]->setParticlePropertyList(P.PropertyList, myIndex);
   }
-  KineticEnergy                      = H[0]->getValue();
-  P.PropertyList[WP::LOCALENERGY]    = LocalEnergy;
-  P.PropertyList[WP::LOCALPOTENTIAL] = LocalEnergy - KineticEnergy;
-  // auxHevaluate(P);
+  updateKinetic(*this, P);
```

Comment:
I wish there was a unit test here and elsewhere to check for instance that the LocalEnergy value is indeed the same, but this refactor looks ok by instpection.

Reply:
Did you mean testing
`P.PropertyList[WP::LOCALENERGY]   == ham.getLocalEnergy()` ?

Reply:
I added checks in one of the unit test.